### PR TITLE
launch configuration from gutter

### LINKED ISSUE #28: Launch run-configuration from gutter
# What would you like to be added
add an icon (green arrow  ) in the gutter to 

- [x] eval the rule on the line 
- [x] eval package  ref #9 
- [x] test rule on the line 
- [x] test a package
- [x] edit the configuration if it already exist

# Why is this needed
To generate configuration fast. Moreover it simpler to the user who doesn't know OPA command line (eg how to run only one test "method"
 

# Implementation details
This icon in the gutter will create the appropriate run configuration of type (`OpaEvalRunConfiguration` or  `OpaTestlRunConfiguration` if not already exist and execute it.
 
This icon in the gutter should  be only available on rego test file

We need to implement a `RunLineMarkerContributor` (ie the green arrow) and `LazyRunConfigurationProducer` to create/run the configuration for each type (eval and test)

### PR DESCRIPTION
**Pull request by vgramer**
# Description
create LineRunMaker(green arrow)  on package and RuleHead. And 2 run configuration producer (one for eval and one for test) to create/edit the associated configuration

Most of the logic has been coded but some corner cases must still be addressed and tests must be added.
 

Fixes #28 #9 

# Special notes for your reviewer
It still in progress but if you want to test it. you must either comment or adapt these lines 
[OpaEvalRunConfigurationProducer.kt#L65-L66](https://github.com/vgramer/opa-idea-plugin/blob/4acdb35f6d80f5fd8b6d0a32aff715357982ba33/src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/OpaEvalRunConfigurationProducer.kt#L65-L66)
[OpaTestRunConfigurationProducer.kt#L59](https://github.com/vgramer/opa-idea-plugin/blob/4acdb35f6d80f5fd8b6d0a32aff715357982ba33/src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/test/producer/OpaTestRunConfigurationProducer.kt#L59)

### CODE REVIEW

File: docs/devel/architecture.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 - [Project structure](#project-structure)
 - [Lexer](#lexer)
 - [Parser](#parser)
+- [Useful methods](#Useful methods)
```
Comment (irodzik) [inline docs/devel/architecture.md:6]: Looks like this link is not working

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/OpaEvalRunConfigurationProducer.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        val (query, options) = queryAndOptions(element, regoPackage)
+
+        configuration.query = query
+        configuration.additionalArgs = "-f pretty $options"
+        configuration.name = configuration.query!!
+        return true
+    }
+
+    /**
```
Comment (lkadalski) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/OpaEvalRunConfigurationProducer.kt:55]: JavaDoc shouldn't be rather on public methods than on private ones ?

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/test/producer/OpaTestRunConfigurationProducer.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+        configuration.additionalArgs = "-f pretty  ${getRunOption(element, regoPackage)}"
+        configuration.name = element.text
+        return true
+    }
+
+
+    /**
+     * generate the run option to pass to opa eval. This option allow to filter the test to execute by passing a regex
```
Comment (lkadalski) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/test/producer/OpaTestRunConfigurationProducer.kt:53]: same here

File: src/test/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/RunConfigurationProducerTestBase.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    protected fun checkOnLeaf() = checkOnElement<PsiElement>()
+
+    private inline fun <reified T : PsiElement> checkOnElement() {
+        val configurationContext = ConfigurationContext(
+            myFixture.file.findElementAt(myFixture.caretOffset)?.ancestorOrSelf<T>()
+        )
+        check(configurationContext)
+    }
+
+    //TODO move
```
Comment (lkadalski) [inline src/test/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/RunConfigurationProducerTestBase.kt:30]: What about this?

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+// see com.intellij.codeInsight.daemon.LineMarkerProvider
+class OpaCommandRunLineMarker : RunLineMarkerContributor() {
+    override fun getInfo(element: PsiElement): Info? {
+        if (element.containingFile !is RegoFile) return null
+        if (element.elementType != RegoTypes.ASCII_LETTER) return null
+
+        val cmd = if (element.containingFile.name.startsWith("test_")) "test" else "eval"
+
+        // we do not use PsiTreeUtil.getParentOfType() of type because for performance reason marker must be placed on leaf
```
Comment (lkadalski) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt:26]: maybe it should be in JavaDoc?

File: src/test/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/RunConfigurationProducerTestBase.kt (hunk 4 above)
Comment (vgramer) [reply src/test/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/RunConfigurationProducerTestBase.kt:30]: nice catch thanks

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/OpaEvalRunConfigurationProducer.kt (hunk 2 above)
Comment (vgramer) [reply src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/producer/OpaEvalRunConfigurationProducer.kt:55]: @lkadalski it was to put a meaning (have an example ) on `"data.${regoPackage?.ref?.text}.${element.text}"`.  I have used Javadoc because it allows to quickly know what kind of query is generated (thanks to `quick documentation` Idea feature) when you read the code line 32 or 47.

 Do you prefer I put It in a comment inside the function?

File: docs/devel/architecture.md (hunk 1 above)
Comment (vgramer) [reply docs/devel/architecture.md]: done

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * Note: for performance reason marker must be placed on a leaf
+     */
+    override fun getInfo(element: PsiElement): Info? {
+        if (element.containingFile !is RegoFile) return null
+        if (element.elementType != RegoTypes.ASCII_LETTER) return null
+
+        val cmd = if (element.containingFile.isRegoTestFile()) "test" else "eval"
+
+        // we do not use PsiTreeUtil.getParentOfType() because marker must be placed on leaf(in our case ASCII_LETTER)
```
Comment (vgramer) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt:39]: @lkadalski is that what you mean in your previous comment?  or do you prefer I add an implementation detail section with the explanation of `PsiTreeUtil.getParentOfType()` in the Javadoc ?

Comment (lkadalski) [reply src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt:39]: Example should also be in javadoc above method. I think that line comments are generally not welcome as I see open source projects (like Spring) ;)

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import com.intellij.icons.AllIcons
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.elementType
+import org.openpolicyagent.ideaplugin.lang.psi.*
+
+// see com.intellij.codeInsight.daemon.LineMarkerProvider
+/**
+ * Add the marker in the gutter (ie the green arrow) to generate / run  RunConfiguration
+ *
+ * The RunConfiguration are generated / run by [org.openpolicyagent.ideaplugin.ide.runconfig.producer.OpaEvalRunConfigurationProducer]
```
Comment (lkadalski) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt:19]: :+1:

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+package org.openpolicyagent.ideaplugin.ide.linemarkers
+
+import com.intellij.execution.lineMarker.ExecutorAction
+import com.intellij.execution.lineMarker.RunLineMarkerContributor
+import com.intellij.icons.AllIcons
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.elementType
+import org.openpolicyagent.ideaplugin.lang.psi.*
+
+// see com.intellij.codeInsight.daemon.LineMarkerProvider
```
Comment (lkadalski) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/linemarkers/OpaCommandRunLineMarker.kt:15]: You can move this to javadcoc

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/test/producer/OpaTestRunConfigurationProducer.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        return configuration.additionalArgs?.matches(Regex(".*${getRunOption(element, regoPackage)}( .*|\$)")) ?: false
+    }
+
+    override fun setupConfigurationFromContext(
+        configuration: OpaTestRunConfiguration,
+        context: ConfigurationContext,
+        sourceElement: Ref<PsiElement>
+    ): Boolean {
+        val element = sourceElement.get()
+        if (! element.containingFile.isRegoTestFile()) return false
```
[the remark this answers is not in the document]
Comment (asadali) [reply src/main/kotlin/org/openpolicyagent/ideaplugin/ide/runconfig/test/producer/OpaTestRunConfigurationProducer.kt:42]: agreed on the suggestion to ignore filename. any rule with `test_` prefix will work for opa test.
we can improve it in a follow-up PR.
